### PR TITLE
Layer Cluster Generation

### LINKED ISSUE #21: Generate usage layers for all groups, software, and mitigations
As a user, I want to be able to easily create usage layers (as implemented in #13/#9) for all of the groups, software, or mitigations in the dataset. 

While reasonably easy to run the existing UsageGenerator over a series of object names, building that feature into the library would save effort for users who want to build a full set of layers representing for the ATT&CK knowledge base.

This improvement should include:
- An API that returns a set of Layer objects (either in an array, or in a map of `name/ATT&CK-ID/STIX-ID -> Layer`)
- A CLI which writes the set of layer files into a folder
- Both methods should take the object type (group, software, mitigation) as a parameter.

### PR DESCRIPTION
**Pull request by ArtificialErmine**
Adds the ability to generate sets of layers automatically. 

Closes #21

### CODE REVIEW

**Comment by ArtificialErmine** [comment]
This instance of it is rather slow on the extreme end (groups in enterprise can run about 2 hours according to the estimate), but unfortunately I haven't found any ways to quickly speed it up (there doesn't seem to be any obvious places we can cache results for the moment). If you target an output that is much more limited, like groups for mobile, the output gets generated at reasonable speeds.

**Review by isaisabel — CHANGES_REQUESTED** [review]
The general approach looks solid and I like the fact that it's just batch running existing code, very clean and simple approach. Most of my comments are regarding documentation/naming of functionality for clarity of the user. 

In terms of performance improvements, [this block](https://github.com/mitre-attack/mitreattack-python/blob/master/mitreattack/navlayers/generators/usage_generator.py#L65-L68) can instead of running a filter function iterate over the mapped IDs and use a `self.source_handle.get(id)` to actually get the mapped object. I believe [stix2's `get` function runs in constant time](https://github.com/oasis-open/cti-python-stix2/blob/master/stix2/datastore/memory.py#L260-L274) which is better than the Filter function which must iterate the entire contents of the data source.

Pseudocode:
```python
results = []
for r in related:
    mapped = self.source_handle.get(r.related)
    if mapped.type != 'attack-pattern': continue
    results.append(mapped)
return remove_revoked_deprecated(results), obj
```

Efficiency comparison (where N is the number of objects in the dataset and the majority of objects are relationships):
- Before: All Objects * Relevant Relationships * All Objects = `O(N^3)`  
- After: All Objects * Relevant Relationships = `O(N^2)`

(This particular big-O analysis is a bit tricky so that my be a bit off, but I think it's something along that line. At the very least it will be reducing one iteration of the full dataset per output layer)

File: mitreattack/navlayers/generators/sum_generator.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        :param local: Optional path to local data
+        """
+        self.usage_handle = UsageLayerGenerator(source, domain, local)
+        self.mapping = dict(group=[Filter('type', '=', 'intrusion-set')],
+                            software=[Filter('type', '=', 'malware'), Filter('type', '=', 'tool')],
+                            mitigation=[Filter('type', '=', 'course-of-action')])
+
+    def generate_layers(self, layers_type):
+        """
+        Generate a layer
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/generators/sum_generator.py:23]: Yeah, that documentation was definitely off.

File: mitreattack/navlayers/generators/sum_generator.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from stix2 import Filter, datastore
+from mitreattack.navlayers.generators.usage_generator import UsageLayerGenerator
+from mitreattack.navlayers.core.exceptions import typeChecker, categoryChecker
+from mitreattack.navlayers.generators.gen_helpers import remove_revoked_depreciated, get_attack_id
+from tqdm import tqdm
+
+
+class SumGenerator:
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/generators/sum_generator.py:8]: Batch makes sense to me - I'll adjust things appropriately.

File: mitreattack/navlayers/layerGenerator_cli.py [hunk 3]
```diff
@@ -11,7 +13,12 @@ def main():
                        help='Output a matrix where the target type is summarized across the entire dataset.')
     group.add_argument('--mapped-to', help='Output techniques mapped to the given group, software, or mitigation. '
                                            'Argument can be name, associated group/software, or ATT&CK ID.')
-    parser.add_argument('-o', '--output', help='Path to the output layer file', default='generated_layer.json')
+    group.add_argument('--mass-type', choices=['group', 'software', 'mitigation'],
+                       help='Output a collection of matrices to the specified folder, each one representing a '
+                            'different instance of the target type.')
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/layerGenerator_cli.py:18]: Batch definitely sounds better. I'll adjust everything to that, and rename things to properly reference layers vs matrices.

File: mitreattack/navlayers/layerGenerator_cli.py [hunk 4]
```diff
@@ -11,7 +13,12 @@ def main():
...
+    parser.add_argument('-p', '--path', help='Path to the output layer directory (mass-type)',
+                        default='generated_layers')
+    parser.add_argument('-o', '--output', help='Path to the output layer file (output)', default='generated_layer.json')
```
[the remark this answers is not in the document]
Comment (ArtificialErmine) [reply mitreattack/navlayers/layerGenerator_cli.py:21]: That would be a lot cleaner... I'll see if I can put something together.

**Comment by ArtificialErmine** [comment]
> The general approach looks solid and I like the fact that it's just batch running existing code, very clean and simple approach. Most of my comments are regarding documentation/naming of functionality for clarity of the user.
> 
> In terms of performance improvements, [this block](https://github.com/mitre-attack/mitreattack-python/blob/master/mitreattack/navlayers/generators/usage_generator.py#L65-L68) can instead of running a filter function iterate over the mapped IDs and use a `self.source_handle.get(id)` to actually get the mapped object. I believe [stix2's `get` function runs in constant time](https://github.com/oasis-open/cti-python-stix2/blob/master/stix2/datastore/memory.py#L260-L274) which is better than the Filter function which must iterate the entire contents of the data source.
> 
> Pseudocode:
> 
> ```python
> results = []
> for r in related:
>     mapped = self.source_handle.get(r.target_ref)
>     if mapped.type != 'attack-pattern': continue
>     results.append(mapped)
> return remove_revoked_deprecated(results), obj
> ```
> 
> Efficiency comparison (where N is the number of objects in the dataset and the majority of objects are relationships):
> 
> * Before: All Objects * Relevant Relationships * All Objects = `O(N^3)`
> * After: All Objects * Relevant Relationships = `O(N^2)`
> 
> (This particular big-O analysis is a bit tricky so that my be a bit off, but I think it's something along that line. At the very least it will be reducing one iteration of the full dataset per output layer)

So, I tested both approaches with a job that takes about half an hour, and unfortunately, we seem to be at the point where the constant is large enough that the complexity shift doesn't actually make a noticeable difference. In a very rough test, it took just under a minute longer to use `get` than to use filters. I'm going to keep toying with this, but I don't think using `get` is quite as beneficial as we might have hoped.

**Review by isaisabel — APPROVED** [review]
Sad that we couldn't get it to be more efficient but I understand that with the amount of data we're chugging through that may be unavoidable. Overall seems fine as it is currently.

**Pull request merged**